augur/io/metadata.py

```diff
@@ -88,3 +97,228 @@ def read_metadata(metadata_file, id_columns=("strain", "name"), chunk_size=None)
         **kwargs
     )
+
+
+def read_metadata_to_dict(metadata, duplicate_reporting=DataErrorMethod.ERROR, id_column=None):
+    """
+    Read rows from *metadata* file and yield each row as a single dict.
+
+    Will report duplicate records based on the *id_column* if requested via
+    *duplicate_reporting* after the generator has been exhausted.
+
+    Parameters
+    ----------
+    metadata: str
+        Path to a CSV or TSV metadata file or IO buffer
+
+    duplicate_reporting: DataErrorMethod, optional
+        How should duplicate records be reported
+
+    id_column: str, optional
+        Name of the column that contains the record identifier used for reporting duplicates.
+        Uses the first column of the metadata if not provided.
+
+    Yields
+    ------
+    dict:
+        The parsed row as a single record
+
+    Raises
+    ------
+    AugurError:
+        When there are parsing errors from the csv standard library or when
+        requested to error on duplicate records
+    """
+    valid_delimiters = [',', '\t']
+    seen_ids = set()
+    duplicate_ids = set()
+    with open_file(metadata) as handle:
+        # Get sample to determine delimiter
+        metadata_sample = handle.read(1024)
+
+        if handle.seekable():
+            handle.seek(0)
+        else:
+            metadata_sample_file = StringIO(metadata_sample)
+            handle = chain(metadata_sample_file, handle)
+
+        try:
+            dialect = csv.Sniffer().sniff(metadata_sample, valid_delimiters)
+        except csv.Error as err:
+            raise AugurError(
+                "Could not determine the delimiter of the metadata file. "
+                "Metadata file must be a CSV or TSV file."
+            ) from err
+
+        metadata_reader = csv.DictReader(handle, dialect=dialect)
+        if duplicate_reporting is DataErrorMethod.SILENT:
+            # Directly yield from metadata reader since we do not need to check for duplicate ids
+            yield from metadata_reader
+        else:
+            if id_column is None:
+                id_column = metadata_reader.fieldnames[0]
+
+            for record in metadata_reader:
+                record_id = record.get(id_column)
+                if record_id is None:
+                    raise AugurError(f"The provided id column {id_column!r} does not exist in the metadata.")
+
+                if record_id in seen_ids:
+                    duplicate_ids.add(record_id)
+                else:
+                    seen_ids.add(record_id)
+
+                yield record
+
+    if duplicate_reporting is not DataErrorMethod.SILENT and duplicate_ids:
+        duplicates_message = f"The following records are duplicated in {metadata!r}:\n" + "\n".join(sorted(duplicate_ids))
+
+        if duplicate_reporting is DataErrorMethod.WARN:
+            print_err(f"WARNING: {duplicates_message}")
+        elif duplicate_reporting is DataErrorMethod.ERROR:
+            raise AugurError(duplicates_message)
```

Comment:
I still need to read through how this is used, but a new `class DuplicateError(AugurError)` might be appropriate, similar to how there is `class FilterException(AugurError)`.

Reply:
I'm curious for your thoughts on when it's appropriate to define a new error class. Why do we have a specific `FilterException` when all other errors are just `AugurError`?

Reply:
(Victor should answer too, but I'll chime in with my 2¢ in the meantime: add a new error class when we have reason (either internally or externally) to catch and handle that error condition _specifically_. If we're not going to catch it specifically, then it can be a general grab bag error like `AugurError` or similar.)

Reply:
Whoops, I missed these replies! What @tsibley said sounds reasonable. With that, `FilterException` is not handled specifically so it can be removed: #1093